com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/UploadAppStep.java

```diff
@@ -81,17 +85,45 @@ public StepPhase executeAsyncStep(ExecutionWrapper execution) throws FileStorage
         return StepPhase.POLL;
     }
 
+    private ApplicationResources getApplicationResources(ExecutionWrapper execution, CloudControllerClient client, CloudApplication app,
+        String appArchiveId,
+        String fileName)
+        throws FileStorageException {
+        ApplicationResources applicationResources = new ApplicationResources();
+        final DelegateExecution context = execution.getContext();
+        FileDownloadProcessor calculateDigestOfApplication = new DefaultFileDownloadProcessor(StepsUtil.getSpaceId(context),
+            appArchiveId, appArchiveStream -> {
+                long maxSize = configuration.getMaxResourceFileSize();
+                try {
+                    ApplicationArchiveExtractor appExtractor = getApplicationArchiveExtractor(appArchiveStream, fileName, maxSize,
+                        Collections.emptySet());
+                    applicationResources.createCloudResources(appExtractor.getApplicationMetaData());
+                    String newApplicationDigest = appExtractor.getApplicationDigest();
+                    detectApplicationFileDigestChanges(execution, app, newApplicationDigest, client);
+                } catch (CloudOperationException e) {
+                    throw e;
+                }
+            });
+
+        fileService.processFileContent(calculateDigestOfApplication);
+
+        return applicationResources;
+    }
+
     private UploadToken asyncUploadFiles(ExecutionWrapper execution, CloudControllerClient client, CloudApplication app,
-        String appArchiveId, String fileName) throws FileStorageException {
+        String appArchiveId, String fileName, ApplicationResources applicationResources) throws FileStorageException {
         UploadToken uploadToken = new UploadToken();
         final DelegateExecution context = execution.getContext();
         FileDownloadProcessor uploadFileToControllerProcessor = new DefaultFileDownloadProcessor(StepsUtil.getSpaceId(context),
             appArchiveId, appArchiveStream -> {
                 Path filePath = null;
                 long maxSize = configuration.getMaxResourceFileSize();
                 try {
-                    filePath = extractFromMtar(appArchiveStream, fileName, maxSize);
-                    upload(execution, client, app, filePath, uploadToken);
+                    CloudResources knownRemoteResources = applicationResources.getKnownRemoteResources(client);

```

Comment:
Since we rely to get knownRemoteResources before upload, we have to verify this with unit tests